### DISCUSSION TITLE
Customise response validation errors

### DISCUSSION THREAD

**Discussion by hrishikesh-k** [fastify/fastify#4740] [Q&A] [answered]
Consider this response:

```js
reply.send({
  message: 'hello world'
})
```

with this schema:

```js
response: {
  200: {
    properties: {
      name: {
        type: 'string'
      }
    },
    required: [
      'name'
    ],
    type: 'object'
  }
}
```

It gives me an response with status 500:

```json
{
    "statusCode": 500,
    "code": "FST_ERR_BAD_STATUS_CODE",
    "error": "Internal Server Error",
    "message": "Called reply with an invalid status code: undefined"
}
```

I have a custom error handler set using `fastify.setErrorHandler()` like:

```js
errorHandlerReply.status(errorHandlerError.statusCode).send({
  message: errorHandlerError.message,
  request_id: errorHandlerRequest.awsLambda.context.awsRequestId
})
```

I have created my own class (`ApiError`) of errors which will automatically sent the default error code to 500 if none specified. Whenever I need to throw an error, I use `throw new ApiError()` which works perfectly with my setup. However, I am not sure how to handle or throw a custom error when response validation fails.

**Comment by metcoder95** [comment]
Hey!

It seems that over this line `errorHandlerReply.status(errorHandlerError.statusCode)`, the `errorHandlerError.statusCode` is `undefined`. Which causes Fastify to throw as `undefined` and cannot be used as an HTTP error.
Did you take a look at that already?

**Reply by hrishikesh-k** [reply]
Thanks! 

I'm not asking about the error message, I understand I can fix it by `errorHandlerError.statusCode || 500` in my code. My question is different. I'm interested to know, how can I throw a custom error?

In other words, I am interested to know, how can I perform response schema validation and handle errors myself? For request schema validation, I'm able to do something like:

```ts
preHandler: (preHandlerRequest, preHandlerReply, preHandlerDone) => {
  if (preHandlerRequest.validationError) {
    preHandlerReply.status(400).send({
      errors: preHandlerRequest.validationError.validation.map((validationError : {
        message : string
      }) => {
        return validationError.message
      }),
      message: `${preHandlerRequest.url} contains invalid data`,
      request_id: preHandlerRequest.awsLambda.context.awsRequestId
    })
  }
}
```

But, how to do that for response schema?

**Reply by metcoder95** [reply, marked as answer]
Oh, sorry, I misunderstood the question :) 
You can provide a custom serializer using `SerializerCompiler` that can do that for you, and you can wrap your errors around custom errors. See: https://www.fastify.io/docs/latest/Reference/Validation-and-Serialization/#serializer-compiler

I'm not sure of your use case, but most of the time the custom behaviour from Fastify is enough, as it strips the values that doesn't fit in the Response schema 🙂

**Reply by hrishikesh-k** [reply]
Thanks, I'll try working with that!

**Reply by javierguzman** [reply]
Hello @Hrishikesh-K by any chance did you achieve what you were looking for? I am trying to do something very similar to you and I do not get it. I would appreciate any hint. Thank you in advance and regards

**Reply by hrishikesh-k** [reply]
I did not consider pursuing that alternative after trying it a bit. The way I was implementing it was by validating the response against the schema myself and sending the response accordingly. I soon found that was very suboptimal for the gains I was getting, especially since response is something I have control over, so I just decided to make my response as schema-complaint as possible.

Sorry if it's not too useful.

**Reply by javierguzman** [reply]
Thank you for replying! So basically for validation errors you use the default Fastify response and for other errors you use your own?

**Reply by hrishikesh-k** [reply]
Yes, to validate response schema, I used Fastify and my custom errors were always handled separately. I have a custom error handler set for my instance, and a custom ApiError class that formats the error as I'd prefer.

**Reply by javierguzman** [reply]
Thanks! I have achieved what I want it. I place the code here just in case it helps future people. Not sure if it is the best way to do it but I think it works:

```
export const errorHandler = (
  err: FastifyError,
  request: FastifyRequest,
  reply: FastifyReply
) => {
  if (err instanceof ApiError) {
    return reply.status(err.statusCode).send({ error: err.message });
  } else if (err.validation) {
    // extract  messages from array of validation errors coming from Fastify
    let fieldName;
    let fieldError;
    const errorMessages = err.validation.map(validationError => {
      fieldName = validationError.instancePath.split('/')[1];
      fieldError = validationError.message;
      return fieldName + ' ' + fieldError;
    });
    return reply.code(400).send({ error: errorMessages });
  } else {
    return reply.send({ error: err.message });
  }
};

export const notFoundErrorHandler = (
  request: FastifyRequest,
  reply: FastifyReply
) => {
  const notFoundPath = request.raw.url;
  return reply.code(404).send({ error: `${notFoundPath} not found` });
};
```

**Reply by Eomm** [reply]
Old but it should still work: https://github.com/fastify/example/tree/master/validation-messages

**Reply by javierguzman** [reply]
Thanks @Eomm! I did not know about that though it seems is not very flexible in terms of the message. If I understood well, you would have to specify all fields, and for example, you would not know if the field is missing or it does not comply with a pattern and stuff like that. Is my understanding correct?

**Reply by Eomm** [reply]
Nope, you can configure a lot of stuff by configuring your json schema like this:
https://ajv.js.org/packages/ajv-errors.html

**Reply by javierguzman** [reply]
Oh dear...I think that is too fine detail for me, maybe in future I will give it a play because it seems very useful, mostly for i18n. For my use case, I think is easier what I have done.
